h2o-py/h2o/frame.py

```diff
@@ -1960,34 +1961,45 @@ def as_data_frame(self, use_pandas=True, header=True, multi_thread=False):
         if can_use_pandas() and use_pandas:
             import pandas
             if multi_thread:
-                if can_use_datatable():
-                    try:
+                have_correct_toolBox = can_use_datatable() or (can_use_polars() and can_use_pyarrow())
+                if have_correct_toolBox:
+                    try: 
                         tmpdir = tempfile.mkdtemp()
                         fileName = os.path.join(tmpdir, "h2oframe2Convert.csv")
                         h2o.export_file(self, fileName)
-                        #h2o.download_csv(self, fileName)
-                        import datatable as dt
-                        frameTypes = self.types
-                        validFrameTypes = {}
-                        for key, value in frameTypes.items():
-                            if value.startswith('int'):
-                                validFrameTypes[key] = dt.int64
-                            elif value.startswith("real"):
-                                validFrameTypes[key] = dt.float64
-                        dt_frame = dt.fread(fileName, na_strings=[""], columns=validFrameTypes)
-                        return dt_frame.to_pandas()
+                        if can_use_datatable() and not(module == 'polars'): # default to datatable unless specified
+                            return self.convert_with_datatable(fileName)
+                        elif can_use_polars() and can_use_pyarrow():  # 
```

Comment:
This seems incorrect to me. You require the `pyarrow` and yet you don't use it with the polars.

AFAIK there are 4 scenarios that can occur here:

- using pyarrow to load the csv and then convert it to pandas (pyarrow internal datastructure is supported in recent pandas versions so no copy occurs here)
- using datatable (we have that)
- using polars (this is what you added in this PR)
- using polars and pyarrow to avoid copying and reformatting data when converting from polars to pandas

I added the code that I used to benchmark that along with the results (in seconds).

```python
  with MeasureTime("h2o -> csv -> pyArrow -> pandas"):
      _, tmp = tempfile.mkstemp(suffix=".csv")
      os.unlink(tmp)
      try:
          h2o.export_file(frame, tmp)
          dt_frame = pyarrow.csv.read_csv(tmp)
          pd_frame = dt_frame.to_pandas()
      finally:
          os.unlink(tmp)
      
   with MeasureTime("h2o -> csv -> datatable -> pandas"):
      _, tmp = tempfile.mkstemp(suffix=".csv")
      os.unlink(tmp)
      try:
          h2o.export_file(frame, tmp)
          dt_frame = dt.fread(tmp)
          pd_frame = dt_frame.to_pandas()
      finally:
          os.unlink(tmp)
          
  with MeasureTime("h2o -> csv -> polars -> pandas"):
      _, tmp = tempfile.mkstemp(suffix=".csv")
      os.unlink(tmp)
      try:
          h2o.export_file(frame, tmp)
          dt_frame = pl.read_csv(tmp)
          pd_frame = dt_frame.to_pandas()
      finally:
          os.unlink(tmp)
          
  with MeasureTime("h2o -> csv -> polars (pyArrow) -> pandas"):
      _, tmp = tempfile.mkstemp(suffix=".csv")
      os.unlink(tmp)
      try:
          h2o.export_file(frame, tmp)
          dt_frame = pl.read_csv(tmp)
          pd_frame = dt_frame.to_pandas(use_pyarrow_extension_array=True)
      finally:
          os.unlink(tmp)       
```

![Screenshot 2023-11-22 at 9 52 51](https://github.com/h2oai/h2o-3/assets/61695433/d7f34259-e394-4d93-ab5e-9f3a53151384)

Reply:
@tomasfryda 

Polars uses pyarrow to do the to_pandas conversion.  Hence, I need to have pyarrow.  Here is the link: https://pola-rs.github.io/polars/py-polars/html/reference/dataframe/api/polars.DataFrame.to_pandas.html

I tried to just use your pyarrow suggestion.  Python 3.6 works fine but Python 3.11 is not happy.  I get this error:

<img width="1056" alt="image" src="https://github.com/h2oai/h2o-3/assets/7231712/fb8a80a1-fb0c-4798-9796-c57ca2c2591a">

It just means that the null value is empty string instead of float nan

Reply:
Oh, sorry I thought pyarrrow is used just when using the `use_pyarrow_extension_array`. But you are right.

Using `use_pyarrow_extension_array` requires recent pandas and pyarrow. It would be nice to allow user to use it since it can bring significant performance benefit (especially if you don't have enough memory to duplicate the data) but the restrictions on pandas and pyarrow versions are not good for having it as a default.

Reply:
I tried using pyarrow all the way, it was not able to handle NaN when used with Python 3.6.  However, it does not have a problem with Python 3.9 or above.